### PR TITLE
Add support for application/JSON calls in BP2

### PR DESCRIPTION
**Pull request by maheshkelkar**
- Added prefix "Bp" to all the BP errors/exceptions
- Throw Bp exceptions for all the error responses generated by BP2
- Convert exceptions to error responses in the ExceptionFilter
- If request is exepcting an application/json response, then return
  a response with custom response (e.g. contains a redirect_url)

### CODE REVIEW

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala [hunk 1]
```diff
@@ -60,21 +63,22 @@ object Keymaster {
...
-        //  Preserve Response Status code by throwing AccessDenied exceptions
+        case Status.Forbidden => {
+          responseDenied.incr
+          Future.exception(BpIdentityProviderError(Status.Forbidden,
+            s"IdentityProvider denied user: ${req.credential.uniqueId} with status: ${res.status}"))
+        }
         case _ => {
-          log.debug(s"Keymaster IdentityProvider denied user: ${req.credential.uniqueId} " +
-            s"with status: ${res.status}")
           responseFailed.incr
```
Comment (PeterFCaswell) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: Would it be clearer here if the variable was named, for instance "statResponseFailed"? This applies to the other statistic names.

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala [hunk 2]
```diff
@@ -60,21 +63,22 @@ object Keymaster {
           Tokens.derive[Tokens](res.contentString).fold[Future[IdentifyResponse[Tokens]]](
             err => {
               responseParsingFailed.incr
-              Future.exception(IdentityProviderError(Status.InternalServerError,
-                "Failed to parse the Keymaster Identity Response"))
+              Future.exception(BpTokenParsingError("Failed to parse the Keymaster Identity Response"))
```
Comment (PeterFCaswell) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: Can you explain why you've added the 'Bp' prefix? Aren't all these names already qualified with a borderpatrol namespace?

File: core/src/main/scala/com/lookout/borderpatrol/Binder.scala [hunk 3]
```diff
@@ -100,11 +100,9 @@ object BinderBase {
       cl <- getOrCreate(name, urls)
       res <- cl.apply(request)
     } yield res) handle {
-      case e => {
-        log.warning("Failed to connect " +
-          s"for: $name to: ${urls.map(u => u.getAuthority).mkString(",")} with: ${e.getMessage}")
-        throw CommunicationError(s"${name} with ${e.getMessage}")
-      }
+      case e =>
+        throw BpCommunicationError(s"Failed to connect for: $name " +
```
Comment (PeterFCaswell) [inline core/src/main/scala/com/lookout/borderpatrol/Binder.scala]: I would recommend quotes around things like $name. It is a pain in the ass to be reading an error message where parameters like this have spaces in them and you can't easily determine what is the name and what is the error message.

(And no, I don't always remember to do this, and yes, please point it out to me when I miss it. :-) )

File: core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala [hunk 4]
```diff
@@ -370,28 +340,75 @@ case class RewriteFilter() extends SimpleFilter[BorderRequest, Response] {
...
-    case error: IdentityProviderError => warningAndResponse(error.getMessage, error.status)
-    case error: Exception =>
-      log.error(error, error.getMessage)
-      tap(Response(Status.InternalServerError))(
-        r => { r.contentString = error.getMessage; r.contentType = "text/plain" }
-      )
+  def errorHandler(req: Request): PartialFunction[Throwable, Response] = {
+    case error: BpAccessIssuerError => warningAndResponse(req, error.getMessage, error.status)
+    case error: BpIdentityProviderError => warningAndResponse(req, error.getMessage, error.status)
+    case error: BpRedirectError =>
```
Comment (PeterFCaswell) [inline core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala]: A comment here would be famous.

File: core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala [hunk 5]
```diff
@@ -370,28 +340,75 @@ case class RewriteFilter() extends SimpleFilter[BorderRequest, Response] {
...
+              ("redirect_url", error.location.asJson)
+            )).toString()
+            res.contentType = "application/json"
+          case _ =>
+            res.status = Status.Found
+            res.location = error.location
+            res.contentString = error.msg
+            res.contentType = "text/plain"
+        }})
+    case error: BpLogoutError =>
```
Comment (PeterFCaswell) [inline core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala]: A comment here would be much appreciated.

File: core/src/main/scala/com/lookout/borderpatrol/package.scala [hunk 6]
```diff
@@ -56,8 +56,9 @@ package object borderpatrol {
   }
 
   object errors {
-    abstract class BorderError(val status: http.Status, val description: String) extends Exception
-    class InvalidRequest(description: String = "") extends BorderError(http.Status.BadRequest, description)
+    class BorderError(val status: http.Status, val description: String) extends Exception(description)
```
Comment (PeterFCaswell) [inline core/src/main/scala/com/lookout/borderpatrol/package.scala]: If you're prefixing 'Bp' to all your errors, shouldn't these have the same prefix?

File: core/src/main/scala/com/lookout/borderpatrol/sessionx/SignedId.scala [hunk 7]
```diff
@@ -156,7 +156,7 @@ object SignedId {
     implicit def bytes2Secret(bytes: IndexedSeq[Byte])(implicit store: SecretStoreApi): Try[Secret] =
       store.find(_.id == bytes) match {
         case Some(s) => Success(s)
-        case None => Failure(new SignedIdError(s"No secret with id=$bytes"))
+        case None => Failure(new BpSignedIdError(s"No secret with id=$bytes"))
```
Comment (PeterFCaswell) [inline core/src/main/scala/com/lookout/borderpatrol/sessionx/SignedId.scala]: This is another case where delimiters around $bytes would be really helpful to human readers. Especially if the bytes randomly have newlines in them.

**Comment by PeterFCaswell** [comment]
That's all I have for this go-round.

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala (hunk 1 above)
Comment (maheshkelkar) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: ok, will fix it.

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala (hunk 2 above)
Comment (maheshkelkar) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: Almost all the Borderpatrol Error(s) are far too generic (e.g. CommunicationError, SessionError, etc) and hard to differentiate from finagle or java errors/exceptions. Hence I decided to add a `Bp` prefix

File: core/src/main/scala/com/lookout/borderpatrol/Binder.scala (hunk 3 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/Binder.scala]: ok, will do.

File: core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala (hunk 4 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala]: done.

File: core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala (hunk 5 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala]: done :)

File: core/src/main/scala/com/lookout/borderpatrol/package.scala (hunk 6 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/package.scala]: Yes

File: core/src/main/scala/com/lookout/borderpatrol/sessionx/SignedId.scala (hunk 7 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/sessionx/SignedId.scala]: ok

File: core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala [hunk 8]
```diff
@@ -370,28 +339,82 @@ case class RewriteFilter() extends SimpleFilter[BorderRequest, Response] {
...
+          res.status = Status.Found
+          res.location = error.location
+          res.contentString = error.msg
+          res.contentType = "text/plain"
+      }})
+  }
+
+  /* Convert this error into a response appropriate to the client */
+  private[this] def warningAndLogoutResponse(req: Request, error: BpLogoutError): Response = {
+    log.info(error.msg)
```
Comment (PeterFCaswell) [inline core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala]: Why is this at info and the log above is at warning level?

**Comment by PeterFCaswell** [comment]
Just a minor issue. If you say the log level difference is on purpose, I'll merge it.

File: core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala (hunk 8 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala]: I should change the method name.

Logout & Redirect, both should be at info level. I will update it.

**Pull request merged**